File: nodecore-miners-pop/src/main/java/nodecore/miners/pop/Threading.java
// VeriBlock PoP Miner
// Copyright 2017-2019 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.miners.pop;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import org.bitcoinj.utils.ContextPropagatingThreadFactory;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class Threading {
    public static final ExecutorService TASK_POOL;

    static {
        TASK_POOL = Executors.newFixedThreadPool(50,
                new ThreadFactoryBuilder()
                        .setNameFormat("pop-tasks-%d")
                        .setThreadFactory(new ContextPropagatingThreadFactory("pop-tasks"))
                        .build());
    }

    public static void shutdown() {
        TASK_POOL.shutdown();

        try {
            if (!TASK_POOL.awaitTermination(10, TimeUnit.SECONDS)) {
                TASK_POOL.shutdownNow();
            }
        } catch (InterruptedException ex) {
            TASK_POOL.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
}

File: pop-miners/altchain-pop-miner/src/main/kotlin/org/veriblock/miners/pop/util/Threading.kt
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package org.veriblock.miners.pop.util

import com.google.common.util.concurrent.ThreadFactoryBuilder
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import org.veriblock.lite.util.Threading

object Threading {
    val LISTENER_THREAD: ExecutorService = Executors.newSingleThreadExecutor(
        ThreadFactoryBuilder()
            .setNameFormat("event-listener")
            .build()
    )
    val NODECORE_POLL_THREAD: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor(
        ThreadFactoryBuilder()
            .setNameFormat("nc-poll")
            .build()
    )
    val SI_POLL_THREAD: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor(
        ThreadFactoryBuilder()
            .setNameFormat("si-poll")
            .build()
    )
    val MINER_THREAD: ExecutorService = Executors.newSingleThreadExecutor(
        ThreadFactoryBuilder()
            .setNameFormat("miner-thread")
            .build()
    )
    val TASK_POOL: ExecutorService = Executors.newFixedThreadPool(
        50,
        ThreadFactoryBuilder()
            .setNameFormat("pop-tasks-%d")
            .build()
    )

    init {
        Runtime.getRuntime().addShutdownHook(Thread {
            shutdown()
        })
    }

    @Throws(ExecutionException::class, InterruptedException::class)
    fun shutdown() {
        val shutdownTasks = CompletableFuture.allOf(
            CompletableFuture.runAsync { shutdown(Threading.LISTENER_THREAD) },
            CompletableFuture.runAsync { shutdown(Threading.NODECORE_POLL_THREAD) },
            CompletableFuture.runAsync { shutdown(Threading.MINER_THREAD) },
            CompletableFuture.runAsync { shutdown(Threading.TASK_POOL) }
        )
        shutdownTasks.get(20, TimeUnit.SECONDS)
    }

    private fun shutdown(executorService: ExecutorService) {
        executorService.shutdown()

        try {
            if (!executorService.awaitTermination(10, TimeUnit.SECONDS)) {
                executorService.shutdownNow()
            }
        } catch (ex: InterruptedException) {
            executorService.shutdownNow()
            Thread.currentThread().interrupt()
        }

    }
}
